src/producer.rs

```diff
@@ -73,13 +73,15 @@ pub struct ProducerInternal {
     waiting_confirmations: WaiterMap,
     closed: Arc<AtomicBool>,
     accumulator: MessageAccumulator,
+    sender_lock: Mutex<()>,
 }
 
 impl ProducerInternal {
     async fn batch_send(&self) -> Result<(), ProducerPublishError> {
         let mut count = 0;
         let mut messages = Vec::with_capacity(self.batch_size);
 
+        let guard = self.sender_lock.lock().await;
         while count != self.batch_size {
             match self.accumulator.get().await? {
```

Comment:
What do you think about refactoring the `MessageAccumulator::get` so that it returns all available messages? Something like:

```rust
pub(crate) async fn drain(&self) -> Vec<ClientMessage> {
    let mut receiver = self.receiver.lock().await;
    let mut messages = Vec::with_capacity(self.capacity);
    while let Some(message) = receiver.try_recv().ok() {
        messages.push(message);
    }
    if !messages.is_empty() {
        self.message_count
            .fetch_sub(messages.len(), Ordering::Relaxed);
    }
    messages
}
```

That way the `MessageAccumulator`'s `receiver` mutex does the locking work and only needs to be locked once per `ProducerInternal::batch_send`.

Reply:
To avoid contention problem like in #186  make sense also to do this refactor, but i would still use
the batch size here as limit